src/Symfony/Component/Utf8/GenericStringInterface.php

```diff
@@ -0,0 +1,159 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Utf8;
+
+use Symfony\Component\Utf8\Exception\ExceptionInterface;
+
+/**
+ * Represents a generic string implementation.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ * @author Hugo Hamon <hugo.hamon@sensiolabs.com>
+ *
+ * @throws ExceptionInterface
+ * 
+ * @final
+ */
+interface GenericStringInterface extends \IteratorAggregate
+{
+    public function __toString();
+
+    /**
+     * @return int|null
+     */
+    public function indexOf(string $needle, int $offset = 0);
+
+    /**
+     * @return int|null
+     */
+    public function indexOfIgnoreCase(string $needle, int $offset = 0);
+
+    /**
+     * @return int|null
+     */
+    public function lastIndexOf(string $needle, int $offset = 0);
+
+    /**
+     * @return int|null
+     */
+    public function lastIndexOfIgnoreCase(string $needle, int $offset = 0);
+
+    /**
+     * @return static|null
+     */
+    public function substringOf(string $needle, bool $beforeNeedle = false);
+
+    /**
+     * @return static|null
+     */
+    public function substringOfIgnoreCase(string $needle, bool $beforeNeedle = false);
+
+    /**
+     * @return static|null
+     */
+    public function lastSubstringOf(string $needle, bool $beforeNeedle = false);
+
+    /**
+     * @return static|null
+     */
+    public function lastSubstringOfIgnoreCase(string $needle, bool $beforeNeedle = false);
+
+    /**
+     * @return static
+     */
+    public function reverse();
+
+    public function length(): int;
+
+    public function width(bool $ignoreAnsiDecoration = true): int;
```

Comment:
what about default `false`? or move it to a new implementation? this context semantically affects other methods as well right?